components/central-application-gateway/internal/proxy/unauthorizedresponseretrier.go

```diff
@@ -0,0 +1,105 @@
+package proxy
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/kyma-project/kyma/components/central-application-gateway/internal/metadata/model"
+
+	"github.com/kyma-project/kyma/components/central-application-gateway/pkg/apperrors"
+	log "github.com/sirupsen/logrus"
+)
+
+type retrier struct {
+	apiIdentifier            model.APIIdentifier
+	request                  *http.Request
+	requestBodyCopy          io.ReadCloser
+	retried                  bool
+	timeout                  int
+	updateCacheEntryFunction updateCacheEntryFunction
+}
+
+type updateCacheEntryFunction = func(apiIdentifier model.APIIdentifier) (*CacheEntry, apperrors.AppError)
+
+func newUnauthorizedResponseRetrier(apiIdentifier model.APIIdentifier, request *http.Request, requestBodyCopy io.ReadCloser, timeout int, updateCacheEntryFunc updateCacheEntryFunction) *retrier {
+	return &retrier{apiIdentifier: apiIdentifier, request: request, requestBodyCopy: requestBodyCopy, retried: false, timeout: timeout, updateCacheEntryFunction: updateCacheEntryFunc}
+}
+
+func (rr *retrier) RetryIfFailedToAuthorize(r *http.Response) error {
+	if rr.retried {
+		return nil
+	}
+
+	rr.retried = true
+
+	if r.StatusCode == http.StatusForbidden || r.StatusCode == http.StatusUnauthorized {
+		log.Infof("Request from service with name id %s failed with %d status, invalidating proxy and retrying.", rr.apiIdentifier.Service, r.StatusCode)
+
+		retryRes, err := rr.retry()
+		if err != nil {
+			return err
+		}
+
+		if retryRes != nil {
+			if r.Body != nil {
+				r.Body.Close()
+			}
+			*r = *retryRes
+		}
+	}
+
+	return nil
+}
+
+func (rr *retrier) retry() (*http.Response, error) {
+	request, cancel := rr.prepareRequest()
+	defer cancel()
+
+	request.Body = rr.requestBodyCopy
+
+	cacheEntry, err := rr.updateCacheEntryFunction(rr.apiIdentifier)
```

Comment:
updateCacheEntryFunction will really invoke proxy.createCacheEntry which create new proxy with a new transport. This will leave current open connections unclosed.

Reply:
As agreed, we will handle it in a separate PR.